weld/llvm.rs

```diff
@@ -2384,6 +2384,96 @@ impl LlvmGenerator {
+                // keyfunc is an actual SirFunction
+                let out_ty = func.symbol_type(output)?;
+                match *out_ty {
+                    Vector(ref elem_ty) => {
+                        let elem_ll_ty = self.llvm_type(elem_ty)?;
+                        let mut str_args = String::from("");
+                        let param_syms: Vec<_> = keyfunc.params.keys().collect();
+                        if param_syms.len() == 1 {
+                            let param_ty = keyfunc.symbol_type(param_syms[0])?;
+                            if *param_ty == **elem_ty {
+                                let (param_ll_ty, param_ll_sym) = self.llvm_type_and_name(keyfunc, param_syms[0])?;
+                                str_args.push_str(&format!("{}* {}", param_ll_ty, param_ll_sym));
+                            } else {
+                                return weld_err!("Type mismatch: vector:{ } and sort key parameter:{ }",
+                                                 print_type(&**elem_ty), print_type(&*param_ty));
+                            }
+                        }
+
+                        // Check that type of key is scalar
+                        let key_ll_sym;
+                        let key_ll_ty;
+                        let last_block = keyfunc.blocks.last().unwrap();
+                        if let Terminator::ProgramReturn(ref key_sym) = last_block.terminator {
+                            let key_ty = keyfunc.symbol_type(key_sym)?;
+                            if let Scalar(_) = *key_ty {
+                                self.gen_cmp(key_ty)?;
+                                let (key_ll_tyc, key_ll_symc) = self.llvm_type_and_name(keyfunc, key_sym)?;
+                                key_ll_ty = key_ll_tyc;
+                                key_ll_sym = key_ll_symc;
+                            } else {
+                                return weld_err!("Sort key function must have scalar return type");
+                            }
+                        } else {
+                            return weld_err!("Sort key Function must have return type");
+                        }
+
+                        // Generate prelude code
+                        let mut keyfunc_ctx = &mut FunctionContext::new(false);
+                        for (arg, ty) in keyfunc.locals.iter() {
+                            let arg_str = llvm_symbol(&arg);
+                            let ty_str = self.llvm_type(&ty)?;
+                            keyfunc_ctx.add_alloca(&arg_str, &ty_str)?;
+                        }
+                        for b in keyfunc.blocks.iter() {
+                            for s in b.statements.iter() {
+                                self.gen_statement(s, keyfunc, keyfunc_ctx)?
+                            }
+                        }
+
+                        // Add key function and sort prelude code
+                        let mut func_gen = IdGenerator::new("%keyfunc");
```

Comment:
you need a single global function generator, otherwise you'll get duplicated names